source/src/python/PyRosetta/src/pyrosetta/utility/initialization.py

```diff
@@ -0,0 +1,932 @@
+# (c) Questions about this can be addressed to University of Washington CoMotion, email: license@uw.edu.
+
+###############################################################################
+# PyRosetta initialization files
+
+
+__author__ = "Jason C. Klima"
+
+
+import base64
+import collections
+import datetime
+import inspect
+import json
+import os
+import pyrosetta
+import re
+import tempfile
+import warnings
+import zlib
+
+from pprint import pprint
+from pyrosetta.rosetta.core.simple_metrics.composite_metrics import ProtocolSettingsMetric
+
+
+class PyRosettaInitFileParserBase(object):
+    _database_option_name = "in:path:database"
+    _corrections_option_name = "corrections:"
+    _init_file_extension = ".init"
+    _strftime_format = "%Y-%m-%d-%H-%M-%S"
+
+    def get_pyrosetta_build(self):
+        return pyrosetta._version_string()
+
+    def pyrosetta_build_warning(self, original_pyrosetta_build, current_pyrosetta_build):
+        _msg = os.linesep.join(
+            [
+                "The PyRosetta version that generated the '.init' file "
+                + "does not match the current PyRosetta version. Please inspect "
+                + "the PyRosetta input files if you encounter any issues during "
+                + "or after PyRosetta initialization: {0}".format(self.kwargs["output_dir"]),
+                "Original: {0}".format(original_pyrosetta_build),
+                "Current:  {0}".format(current_pyrosetta_build),
+            ]
+        )
+        warnings.warn(_msg, UserWarning, stacklevel=4)
+
+    @property
+    def was_init_called(self):
+        return pyrosetta.rosetta.basic.was_init_called()
+
+    def validate_init_was_called(self):
+        if not self.was_init_called:
+            raise RuntimeError(
+                "PyRosetta must be already initialized to dump a '.init' file. "
+                + "Please run `pyrosetta.init()` with custom options and try again."
+            )
+
+    def validate_init_was_not_called(self):
+        if self.was_init_called:
+            raise RuntimeError(
+                "PyRosetta must not be already initialized to initialize from a '.init' file. "
+                "Please ensure that `pyrosetta.init()` was not already called (e.g., "
+                "if using a Jupyter notebook, please restart the kernel) and try again."
+            )
+
+
+class PyRosettaInitFileSerializer(object):
+    _chunk_size = 64 * 1024 # bytes
+    _compression_level = 9
+    _encoding = "utf-8"
+    _prefix_string = "[PyRosettaInitTextFile]"
+    _prefix_binary = "[PyRosettaInitBinaryFile]"
+    _tag_str = b'Txt'
+    _tag_obj = b'Obj'
+
+    def dump_json(self, obj):
+        return json.dumps(
+            obj,
+            skipkeys=False,
+            ensure_ascii=False,
+            check_circular=True,
+            allow_nan=False,
+            cls=None,
+            indent=None,
+            separators=(",", ":"),
+            default=None,
+            sort_keys=False,
```

Comment:
Do we expect a significantly large output file here? If not then I would recommend to set `sort_keys=True, indent=2` since these make JSON much more readable for human eyes...

Reply:
Yes, this particular `dump_json` method results in large output strings (hence `separators=(",", ":")` for an improved compression ratio), and is just used for converting `list` and `dict` objects into `str` objects (which are then compressed by `zlib` and `base64`). This is used internally for caching text files and subfiles.

Reply:
I see…, - from your example below it looks like compressed data will be in single chunks so it will not be a big deal to sort keys and indent, see https://github.com/RosettaCommons/rosetta/pull/503#discussion_r2319864126 for further comment. Thanks,